trantor/net/TcpConnection.h

```diff
@@ -58,20 +56,49 @@ class TcpConnection
     virtual void shutdown() = 0;
     virtual void forceClose() = 0;
     virtual EventLoop *getLoop() = 0;
+    template <typename T>
+    void setContext(T &&context)
+    {
+        typedef typename std::remove_cv<
+            typename std::remove_reference<T>::type>::type ContextType;
+        _contextPtr = std::make_shared<ContextType>(std::forward<T>(context));
+    }
+    template <int N>
+    void setContext(const char (&str)[N])
+    {
+        _contextPtr = std::make_shared<const char *>(str);
+    }
+    template <typename T>
+    const T &getContext() const
```

Comment:
This does make the user API a bit easier, but an assert is always a sign of a user bug waiting to happen.  Instead of returning a reference to the object (especially as the object may not exist), it should just return the pointer to it directly or `nullptr`, though I'd personally opt for returning a new invocation of the shared_ptr itself (with increased reference count) via a function named `getContextRC` and name the one that returns a potentially null pointer as `getContextOpt` or so.  Personally I utterly hate loose pointers in public API's so would instead opt for `std::optional` if this were C++17, but C++14 is the minimum required here so a loose pointer would do.  I would absolutely not return a reference though as that is an API that implies that it can never be null (and in fact if it is called at release without asserts the compiler is free to optimize the code in a way that it can assume it will not be null, which can make the shared_ptr checks and all just vanish in many cases, which is highly unsafe).

Reply:
@OvermindDL1 I've modified the methods.
For convenience, I keep the template form of the getter